File: app/src/main/java/com/irishinterest/irishinterest/webservice/WebService.java
package com.irishinterest.irishinterest.webservice;

import android.util.ArrayMap;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import io.reactivex.Flowable;
import retrofit2.http.GET;
import retrofit2.http.Query;

public interface WebService {


    // AUTHORS

    @GET("?value=authors&type=getAll&offset=0")
    Flowable<List<Author>> authors();

    @GET("?value=authors&type=searchByName")
    Flowable<List<Author>> authorsSearchByName(
            @Query("searchBy") String name
    );

    @GET("?value=authors&type=byLastNameStartsWith")
    Flowable<List<Author>> authorsByLetter(
            @Query("startsWith") String letter
    );

    @GET("?value=authors&type=count")
    Flowable<Integer> authorsCount();

    @GET("?value=authors&type=abcCount")
    Flowable<List<CountByLetter>> authorsAtoZCount();

    @GET("?value=authors&type=detailsById")
    Flowable<AuthorDetails> authorDetails(
            @Query("authorId") Integer authorId
    );

    @GET("?value=authors&type=detailsById")
    Flowable<Map<String, List<Author>>> authorsByBookIds(
            @Query("ids") String bookIds
    );


    // CATEGORIES

    @GET("?value=categories")
    Flowable<List<Category>> categories();


    // BOOKS

    @GET("?value=books&type=searchByName")
    Flowable<List<Book>> booksSearchByName(
            @Query("searchBy") String name
    );

    @GET("?value=books&type=byAuthorID")
    Flowable<List<Book>> booksByAuthorID(
            @Query("authorID") Integer authorID,
            @Query("offset") Integer offset
    );

    @GET("?value=books&type=byCategory")
    Flowable<List<Book>> booksByCategoryId(
            @Query("authorID") Integer categoryId,
            @Query("offset") Integer offset
    );

    @GET("?value=books&type=getLatest2")
    Flowable<List<Book>> booksLatest(
            @Query("offset") Integer offset
    );

    @GET("?value=books&type=getPublished")
    Flowable<List<Book>> booksPublished(
            @Query("offset") Integer offset
    );

    @GET("?value=books&type=getComingSoon")
    Flowable<List<Book>> booksComingSoon(
            @Query("offset") Integer offset
    );


    // BOOK DETAILS

    @GET("?value=books&type=getById2")
    Flowable<BookDetails> bookDetails(
            @Query("bookId") Integer bookID
    );
}

File: app/src/main/java/com/irishinterest/irishinterest/webservice/WebService.kt
package com.irishinterest.irishinterest.webservice

import retrofit2.http.GET
import io.reactivex.Flowable
import com.irishinterest.irishinterest.webservice.CountByLetter
import com.irishinterest.irishinterest.webservice.AuthorDetails
import com.irishinterest.irishinterest.webservice.BookDetails
import retrofit2.http.Query

interface WebService {
    // AUTHORS
    @GET("request.php?apiKey=testApiKey&value=authors&type=getAll&offset=0")
    fun authors(): Flowable<List<Author>>

    @GET("request.php?apiKey=testApiKey&value=authors&type=searchByName")
    fun authorsSearchByName(
        @Query("searchBy") name: String?
    ): Flowable<List<Author>>

    @GET("request.php?apiKey=testApiKey&value=authors&type=byLastNameStartsWith")
    fun authorsByLetter(
        @Query("startsWith") letter: String?
    ): Flowable<List<Author>>

    @GET("request.php?apiKey=testApiKey&value=authors&type=count")
    fun authorsCount(): Flowable<Int>

    @GET("request.php?apiKey=testApiKey&value=authors&type=abcCount")
    fun authorsAtoZCount(): Flowable<List<CountByLetter>>

    @GET("request.php?apiKey=testApiKey&value=authors&type=detailsById")
    fun authorDetails(
        @Query("authorId") authorId: Int?
    ): Flowable<AuthorDetails>

    @GET("request.php?apiKey=testApiKey&value=authors&type=detailsById")
    fun authorsByBookIds(
        @Query("ids") bookIds: String?
    ): Flowable<Map<String, List<Author>>>

    // CATEGORIES
    @GET("request.php?apiKey=testApiKey&value=categories")
    fun categories(): Flowable<List<Category>>

    // BOOKS
    @GET("request.php?apiKey=testApiKey&value=books&type=searchByName")
    fun booksSearchByName(
        @Query("searchBy") name: String?
    ): Flowable<List<Book>>

    @GET("request.php?apiKey=testApiKey&value=books&type=byAuthorID")
    fun booksByAuthorID(
        @Query("authorID") authorID: Int?,
        @Query("offset") offset: Int?
    ): Flowable<List<Book>>

    @GET("request.php?apiKey=testApiKey&value=books&type=byCategory")
    fun booksByCategoryId(
        @Query("authorID") categoryId: Int?,
        @Query("offset") offset: Int?
    ): Flowable<List<Book>>

    @GET("request.php?apiKey=testApiKey&value=books&type=getLatest2")
    fun booksLatest(
        @Query("offset") offset: Int?
    ): Flowable<List<Book>>

    @GET("request.php?apiKey=testApiKey&value=books&type=getPublished")
    fun booksPublished(
        @Query("offset") offset: Int?
    ): Flowable<List<Book>>

    @GET("request.php?apiKey=testApiKey&value=books&type=getComingSoon")
    fun booksComingSoon(
        @Query("offset") offset: Int?
    ): Flowable<List<Book>>

    // BOOK DETAILS
    @GET("request.php?apiKey=testApiKey&value=books&type=getById2")
    fun bookDetails(
        @Query("bookId") bookID: Int?
    ): Flowable<BookDetails>
}